### PR TITLE
Add max_anchors and max_context_window support to P-EAGLE

### LINKED ISSUE #567: [RFC]: Add random anchor for p-eagle first position
### Motivation.

Peagle uses COD sampling, which predicts the next token for every possible starting token in a sample. Then also predicts the token after that for some percent of the tokens (e.g. 70%). And then predicts the 3rd token for 70% of the selected 70%, etc.

So our attention mask shape is: $(\text{sequence len} * 2.73)^2$, where $2.73 = \sum_{n=0}^4{0.7^n}$ and 4 is number of speculative tokens. What we're proposing is using `num_anchors` as starting points for peagle training because a good chunk of the original sequence length will have loss mask 0. That would give us attention masks of size: $(\text{num anchors} \times 2.73) \times (\text{num anchors} \times 1.73 + \text{sequence len})$ This would be smaller and potentially helps us train with longer sequence length.

### Proposed Change.

Add the `max_anchors` parameter to the P-Eagle config that caps the number of COD chain starting points. When set, `generate_cod_sample_indices` randomly selects `max_anchors` positions from the loss_mask=1 indices at depth 0 instead of using all positions. The attention mask construction reuses the existing
  `select_anchors` method to identify these starting points, then builds an asymmetric mask where queries are the COD-sampled tokens $(\text{max anchors} × 2.73)$ and keys/values include the full original sequence plus the deeper COD tokens $(\text{max anchors} × 1.73 + \text{sequence len})$. This shrinks the attention mask from
 $(\text{sequence len} × 2.73)^2$ to $(\text{num anchors} \times 2.73) \times (\text{num anchors} \times 1.73 + \text{sequence len})$, enabling training with longer sequence lengths under the same memory budget.

### Any Other Things.

_No response_

### PR DESCRIPTION
**Pull request by orestis-z**
Closes #567

## Purpose

Add `max_anchors` and `max_context_window` support to P-EAGLE to cap memory usage during training on long sequences.

P-EAGLE's COD sampling predicts next tokens for every starting position, producing attention masks of size `(seq_len × 2.73)²`. When `max_anchors` is set, `generate_cod_sample_indices` selects a random contiguous window containing at most `max_anchors` valid (loss_mask=1) positions, shrinking the effective attention footprint while preserving context for the selected anchors. `max_context_window` (default 4096) hard-caps the window size to prevent sparse loss masks from reinflating it.

Changes:
- Add `max_anchors` and `max_context_window` fields to `PEagleSpeculatorConfig`
- Wire the new params through `PEagleDraftModel` and `generate_cod_sample_indices`
- Add `--max-context-window` CLI arg; make `--max-anchors` default to `None` (was 256), defaulting per-model (256 for DFlash, unlimited for P-EAGLE)
- Fix DFlash `max_anchors` default from 3072 to 256
- Fix DFlash `max_anchors` falsy-value handling: use explicit `is None` check instead of `or` coercion
- Sparse sequence edge case: Ensured the `max_context_window` clamp fires even when `len(valid_indices) < max_anchors`. Previously, sparse sequences would fall back to the full sequence length, inadvertently causing the exact OOMs the window was designed to prevent.

## Tests

- New integration test `test_varying_max_anchors` parametrized over `[4, 16, None]`
- Integration test `test_max_context_window` with small window size (32) to exercise the cap branch
- Unit tests (`test_peagle_data.py`) directly testing `generate_cod_sample_indices` for the sparse bypass bug, normal cap path, and no-windowing case
- Verifies loss is finite and backward pass succeeds for each setting

## Checklist

I have filled in:

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan/results, such as providing test command and pasting the results.
- [ ] (Optional) The necessary documentation update.
- [x] I (a human) have written or reviewed the code in this pr to the best of my ability.

### CODE REVIEW

File: scripts/train.py [hunk 1]
```diff
@@ -723,8 +723,11 @@ def parse_args():
     parser.add_argument(
         "--max-anchors",
         type=int,
-        default=256,
-        help="Maximum anchor positions for DFlash training (default: 256)",
+        default=None,
+        help=(
+            "Maximum anchor positions for training. "
+            "Defaults to 256 for DFlash, unlimited for P-EAGLE."
```
Comment (shanjiaz) [inline scripts/train.py]: I think we should default to 3072 lol 256 was a little too low sorry that got in.

Comment (orestis-z) [reply scripts/train.py]: thanks, fixed it

File: src/speculators/models/dflash/core.py [hunk 2]
```diff
@@ -156,7 +156,7 @@ def from_training_args(
             transformer_layer_config=verifier_config,
             draft_vocab_size=kwargs["draft_vocab_size"],
             block_size=kwargs.get("block_size", 8),
-            max_anchors=kwargs.get("max_anchors", 3072),
+            max_anchors=kwargs.get("max_anchors") or 3072,
```
[the remark this answers is not in the document]
Comment (orestis-z) [reply src/speculators/models/dflash/core.py]: Fixed — replaced `or 3072` with an explicit `is None` check so that only `None` triggers the default, not other falsy values.

File: src/speculators/models/peagle/data.py [hunk 3]
```diff
@@ -36,8 +47,27 @@ def generate_cod_sample_indices(
...
+            ]
+        else:
+            all_valid_indices = all_valid_indices[start_idx : start_idx + max_anchors]
+
+        sample_indices = [torch.arange(window_start, window_end, device=device)]
+        n_per_depth = [window_end - window_start]
+    else:
+        sample_indices = [torch.arange(seq_length, device=device)]
+        n_per_depth = [seq_length]
+
```
[the remark this answers is not in the document]
Comment (orestis-z) [reply src/speculators/models/peagle/data.py]: Fixed — restructured the logic so `max_context_window` is applied regardless of whether valid anchors exceed `max_anchors`. Also added a unit test (`test_applied_when_anchors_within_limit`) that reproduces the bypass with sparse loss masks and confirms the fix.

File: tests/integration/models/test_model_forward.py [hunk 4]
```diff
@@ -338,6 +338,16 @@ def test_varying_down_sample_ratio(self, down_sample_ratio):
...
 
+    @pytest.mark.parametrize("max_anchors", [4, 16, None])
+    def test_varying_max_anchors(self, max_anchors):
+        model = make_peagle_model(max_anchors=max_anchors)
+        samples = _make_samples([128])
+        batch = make_batch(max_len=MAX_LEN, samples=samples, hidden_size=HIDDEN_SIZE)
+        draft_tokens, loss, metrics = model(**batch)
+
+        assert loss.isfinite()
+        loss.backward()
```
[the remark this answers is not in the document]
Comment (orestis-z) [reply tests/integration/models/test_model_forward.py:358]: Added `test_max_context_window` in the integration tests with `max_context_window=32` to exercise the cap branch. Also added a dedicated unit test file (`tests/unit/models/test_peagle_data.py`) that directly tests `generate_cod_sample_indices` for this scenario.

File: src/speculators/models/peagle/data.py [hunk 5]
```diff
@@ -36,8 +47,30 @@ def generate_cod_sample_indices(
     device = loss_mask.device
     all_valid_indices = torch.where(loss_mask == 1)[0]
 
-    sample_indices = [torch.arange(seq_length, device=device)]
-    n_per_depth = [seq_length]
+    if max_anchors is not None and all_valid_indices.shape[0] > 0:
```
Comment (shanjiaz) [inline src/speculators/models/peagle/data.py:50]: I'm a little bit worried that this bit causing more graph breaks but otherwise this diff looks great!

**Comment by orestis-z** [comment]
ON HOLD, see https://github.com/vllm-project/speculators/issues/567#issuecomment-4769263774
